### DISCUSSION TITLE
Auto derive Linked and/or Related traits impls

### DISCUSSION THREAD

**Discussion by nappa85** [SeaQL/sea-orm#176] [Ideas]
Hello guy,
I was looking at expanded code from `DeriveRelation` proc macro
```rust
#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::user::Entity",
        from = "Column::UserId",
        to = "super::user::Column::Id",
    )]
    User,
}
```
this code correctly generates `RelationTrait` impl for `Relation`, but not `Linked` or `Related` traits impls, as I expected.

Am I expecting wrong?

**Comment by billy1624** [comment]
Yes, this is the intended behaviour

**Comment by tyt2y3** [comment]
Think of it this way, multiple `Relation` form a `Link`.
We will clarify this concept in the docs soon.

**Comment by nappa85** [comment]
Ok, thank you very much, because I still haven't clear how two or more link to the same linked entity can coexist, and how I can select all together

**Comment by tyt2y3** [comment]
Definition

https://github.com/SeaQL/sea-orm/blob/889e9df923a10207f2e9a931716615e739cdaca8/src/tests_cfg/cake.rs#L35-L48

Usage

https://github.com/SeaQL/sea-orm/blob/889e9df923a10207f2e9a931716615e739cdaca8/src/query/join.rs#L262-L274

It's very similar to Relation otherwise, but allows more than 2 joins and several relations between same types.

**Reply by isgj** [reply]
@tyt2y3 could it be said that `Relation` is a subset (maybe not the right term) of `Linked` trait? Meaning that with `Linked` you can express all kind of `Relation` plus more.
In this case why not use only `Linked` trait? Having a macro to auto derive `Linked` would be really helpful. 
`Relation` has the limitation of having only one link between 2 entities, but with `Linked` there wouldn't be such limitation.
Take this ex: https://github.com/SeaQL/sea-orm/discussions/251 , it could be like

```rs
pub enum Relation {
    CharacterPublicInfo,
    #[sea_orm(
        belongs_to = "super::esi_types::Entity",
        from = "Column::ShipTypeId",
        to = "super::esi_types::Column::TypeId",
        on_update = "Restrict",
        on_delete = "Restrict"
    )]
    ShipType,
    #[sea_orm(
        belongs_to = "super::esi_types::Entity",
        from = "Column::WeaponTypeId",
        to = "super::esi_types::Column::TypeId",
        on_update = "Restrict",
        on_delete = "Restrict"
    )]
    WeaponType,
}
```
with `Relation` this doesn't work even though the links are of one single hop (1 join required), but with `Linked` it's fine.

At the moment it is confusing to me to choose one as `Relation` and the other as `Linked`

**Reply by billy1624** [reply]
> could it be said that `Relation` is a subset (maybe not the right term) of `Linked` trait?

I would say [`Related`](https://docs.rs/sea-orm/0.6.0/sea_orm/entity/trait.Related.html) is a subset of `Linked`. `Relation` alone is just a collection of relations.

I guess you are comparing `Related` against `Linked`? Correct me if I'm wrong :P

**Reply by isgj** [reply]
Yep, I meant `Related` :facepalm: 
If `Linked` trait is generated for every relation in `Relation` enum, `RelationTrait` can be omitted (less generated code).

**Reply by billy1624** [reply]
I don't think we can get rid of `Related` because we are using it on [`ModelTrait::find_related`](https://docs.rs/sea-orm/0.6.0/sea_orm/entity/prelude/trait.ModelTrait.html#method.find_related) to define which related Entity to be selected.

**Reply by isgj** [reply]
Maybe not immediately if you want to give the users some time, but I think (IMHO) it will be a  better API if there is only one kind of relation trait (call it `Related` or `Linked`, or ...). And `Linked` seems to cover all kind of relations, probably it will require some changes.
Also there will be one trait to maintain, instead of 2.

**Reply by tyt2y3** [reply]
I think I wrote it somewhere. Basically `Related` is the direct relations on a ER diagram.
`Linked` is how you'd join in the queries, so it's more like a dynamic concept.
Does that make sense?

**Reply by isgj** [reply]
Maybe you have better knowledge of the code and the concepts here but as I see it only with `Related` one can already load related and the query builder knows how to do the joins. It's only limited to 1 level relations (1-->1, 1-->M) and to M2M relations.

> Basically `Related` is the direct relations on a ER diagram.

Still limited as in reality `Entity1` can have more __direct__ relations with `Entity2` and here we go with `Related` for 1 relation and `Linked` for the remaining relations between `Entity1` and `Entity2`, which is confusing.
I don't see any added value by `Related`. This is my opinion

**Reply by billy1624** [reply]
Actually we have `Related` first, then `Linked` was added afterward to handle 2 or more level of relations.

**Comment by tyt2y3** [comment]
Okay so I have think of something:

```rust
#[derive(Related)]
enum Related {
    #[sea_orm(
        to = "Cake.def()",
        via = "Filling.def().rev()"
    )]
    Cake {
        entity: super::cake::Entity,
        junction: super::cake_filling::Relation,
    }
}
```

to expand into

```rust
impl Related<super::cake::Entity> for Entity {
    fn to() -> RelationDef {
        super::cake_filling::Relation::Cake.def()
    }

    fn via() -> Option<RelationDef> {
        Some(super::cake_filling::Relation::Filling.def().rev())
    }
}
```

But it does not seem too intuitive

**Reply by nappa85** [reply]
Always keep full namespace, no magic, and maybe skip `ref()` calls and create a second key like via_rev for those who needs `rev()` call at the end.
E.g,.
```rust
#[derive(Related)]
enum Related {
    #[sea_orm(
        to = "super::cake_filling::Relation::Cake",
        via = "super::cake_filling::Relation::Filling"
    )]
    Cake {
        entity: super::cake::Entity,
        junction: super::cake_filling::Relation,
    }
}
```
Or
```rust
#[derive(Related)]
enum Related {
    #[sea_orm(
        to = "super::cake_filling::Relation::Cake",
        via_ref = "super::cake_filling::Relation::Filling"
    )]
    Cake {
        entity: super::cake::Entity,
        junction: super::cake_filling::Relation,
    }
}
```
